### DISCUSSION TITLE
Sonoff Pow R2 persistent total over unexpected shutdown

### DISCUSSION THREAD

**Discussion by mikhail5555** [arendst/Tasmota#13781] [Q&A] [answered]
Hey,

I am using a Sonoff Pow R2 behind another power switch which I often turn off.
Now I understood from #3689 that its a feature that the total power is only stored in flash at midnight. 
Which leads to the following: (Total power resetting to last saved point on boot).
![143051147-dd4e3f44-56b2-4df3-8e70-516cf67b093a](https://user-images.githubusercontent.com/7824786/143096806-89e06a0a-9456-4620-840e-eaed20c377a5.png)
![143051447-dabc20bb-4442-4241-8bef-3853d2ca4028](https://user-images.githubusercontent.com/7824786/143096823-2a1fac95-1e90-4307-89ea-d92d7cbe2da2.png)
Is there a way to make this total power usage save on either shutdown or some timely trigger?

**Comment by barbudor** [comment, marked as answer]
Hi @Mikhail5555 
The point is not to write to often to the flash in order to avoid wearing it too early.
Also I would say that willingly shutting down an IOT device is not really what is intended. Don't forget that in some circumstances that could trigger Tasmota recovery procedure, erasing all settings.

This being said, command `SaveData` without any parameter will force saving of values in the flash.
You can then use either a rule based on Time or on a Timer to execute the command.
A rule like below will regurlarly save every XX minutes (here, every 60 minutes...):
```
ON Time#Minute|60 DO SaveData ENDON
```

Or if you want to save at specific time of the day, you can use a Timer to trigger at a certain time of the day a rule such as:
```
ON Clock#Timer=1 DO SaveData ENDON
```
Here, assuming Timer 1 as been set up to run a rule at your time of day choice.

A last option which may not be easy would be if you can trigger a button/switch just 2 seconds before the power is cut, you could also handle that in a rule.

Cheers

**Reply by sfromis** [reply]
Energy consumption values will already be saved at an orderly shutdown, but not if you just cut off power, which is not a good plan for energy monitoring. 

In my setup, I base my energy usage data on what's published each TelePeriod (or shorter when PowerDelta kicks in). This means that I in case of hard power loss only lose up to TelePeriod (5 minutes default) of consumption data. 

If I wanted to maintain the local values stored in Tasmota, my plan would not be to wear down flash by many writes, but instead have a bot republish last values when the device boots with lost data.

**Reply by mikhail5555** [reply]
Hey @barbudor, Thanks for your quick reply!

I am aware about the flash wearing out quicker, so i will use it with care. So far I didn't have any issues with shutting it down, and in worst case i have a backup with all the config ready.

I do usually shut down the Sonoff first (It's the power to my 3d printer, so once it finishes printing I completely shut it down), so triggering the SaveData on power state = off would be ideal.

So if I understand it correctly the following would work then?
```
Rule1 ON Power1#state=0 DO SaveData ENDON
```
Edit:
Don't forget to turn it on
```
Rule1 1
```

**Reply by mikhail5555** [reply]
@sfromis Could you please explain what exactly you mean with a 'orderly shutdown'? 
I set the Sonoff Pow R2 output to off before i shutdown the device.
Since I only use my 3D printer so often, I am okay with writing to flash every time it turn off power to the printer.

**Reply by sfromis** [reply]
What I mean by orderly shutdown is to tell Tasmota to shut down, meaning a restart. Of course, you can also get the settings data (including energy consumption values) saved by other means, as suggested already.

**Reply by mikhail5555** [reply]
Unrelated but are you sure this is possible? Since in the Commands wiki I am unable to find any 'shutdown' command

**Reply by sfromis** [reply]
Well, I already wrote "a restart". That is a command which will shutdown first.

**Reply by ascillato** [reply]
> are you sure this is possible? Since in the Commands wiki I am unable to find any 'shutdown' command

`Restart 2`  is shutdown

https://tasmota.github.io/docs/Commands/#restart

**Comment by barbudor** [comment]
> ```
> Rule1 ON Power1#state=0 DO SaveData ENDON
> ```

Very good idea 👍
As far as incoming power is cut only once after you powered off the load, that's correct.
That just won't protect you against unexpected power cut. But if you loose main power during a print, power consumption will be your least worry 😱

**Reply by mikhail5555** [reply]
Haha very true, and I mean, once the power is off, the energy usage is usually done for today. I actually slightly expected this to be default, but I am happy its so easy to add/adjust.

**Reply by barbudor** [reply]
indeed `restart 2` seems to be a shutdown

**Comment by mikhail5555** [comment]
Hmmm i guess i have to mark this as not solved, after some more experiments it still seems as its not being saved to flash.
![image](https://user-images.githubusercontent.com/7824786/145572932-fa14d792-5760-40a3-8d09-8024aff0fdb4.png)
I tried wtih both `Restart 1` and `SaveData`, unless `restart device with configuration saved to flash` doesn't save things like usage numbers.
Do you maybe have any idea @barbudor what is causing this? (I am sure the rule is being triggered, as i tested this and have seen it in the console output).
